feat: mail provider (#12083)
nextcloud/documentation: developer_manual/digging_deeper/groupware/mail_provider.rst

.. _mail-providers:

========================================
Mail Provider Interface
========================================

Nextcloud apps can use, and register the mail provider interface, to provide or consume mail service functionality to and from other apps. 

Access to mail services from another app is provided by the Mail Manager class. Using this class your app can find available mail providers and services. Please see the appropriate section below for more details.

The reverse is also possible, one can register an app to supply mail services to other apps. This is accomplished by implementing a custom Mail Provider class and Mail Service class then registering your provider with the system. Please see the appropriate section below for more details.

.. _mail-provider-terminology:

Terminology
-----------

> For clarificaftion this is a reference of terminology used.

Reviewer: ```suggestion
For clarification this is a reference of terminology used.
```

> 1. *Provider* - this is the app that provides mail services e.g. "BigHostingApp", this can a general app for a specific protocol

Reviewer: ```suggestion
1. *Provider* - this is the app that provides mail services e.g. "BigHostingApp", this can be an app for a specific protocol
```

Author: No worries, I know my spelling is horrible

The mail provider class is the main class that the mail manager uses to retrieve available services from your app. Each mail provider can have multiple mail services configure for a user.

This class needs to implement the `IProvider` interface and have all the required methods defined.

.. code-block:: php

    namespace OCA\BigHostingApp\Provider;

    use OCP\Mail\Provider\IProvider;
    use OCP\Mail\Provider\IService;

    class MailProvider implements IProvider {

        public function id(): string {
            return 'big-hosting-app';
        }

        public function label(): string {
            return 'Big Hosting App';
        }

        public function hasServices(string $userId): bool {
            // app specific code to check for available services
        }

        public function listServices(string $userId): array {
            // app specific code to list all available services
        }

        public function findServiceById(string $userId, string $serviceId): IService | null {
            // app specific code to find a specific services
        }

        public function findServiceByAddress(string $userId, string $address): IService | null {

>             // app specific code to find a with a specific email address

Reviewer: ```suggestion
            // app specific code to find a service with a specific email address
```

.. code-block:: php
    
    namespace OCA\BigHostingApp\Provider;

    use OCP\Mail\Provider\Address;
    use OCP\Mail\Provider\IAddress;
    use OCP\Mail\Provider\IMessage;
    use OCP\Mail\Provider\IMessageSend;
    use OCP\Mail\Provider\IService;
    use OCP\Mail\Provider\Message;

    class MailService implements IService, IMessageSend {

        public function id(): string {
            return '1 or service1 or anything else';
        }

        public function capable(string $value): bool {
            // app specific code to check if a service is capable of perform a specific function e.g. Sending a Message
        }

        public function capabilities(): array {
            // app specific code to retrieve a list of capabilities
        }

        public function getLabel(): string {
            // app specific code to retrieve the label/description/name of the service
        }

        public function getPrimaryAddress(): IAddress {
            // app specific code to retrieve the primary email address of the service
        }

        public function getSecondaryAddresses(): array {

>             // app specific code to retrieve the secondary (aliases) email addresses of the service

Reviewer: ```suggestion
            // app specific code to retrieve the secondary email addresses (aliases) of the service
```
